feature/projects-BJS2-96054

Deathroels opened the pull request.

Deathroels committed:
BJS2-96054:
-add controller with endpoints
-add body of facade and service

Deathroels committed:
BJS2-96054:
-add dto classes

Deathroels committed:
BJS2-96054:
-fix packages

Deathroels committed:
BJS2-96054:
-add ProjectFilterDto
-changed endpoint method signature

Deathroels committed:
BJS2-96054:
-add fields and validate ProjectFilterDto, ProjectUpdateDto, ProjectCreateDto
-add fields ProjectDto
-add ProjectMapper
-changed endpoint method signature

Deathroels committed:
BJS2-96054:
-added and implemented the project creation method

Deathroels committed:
BJS2-96054:
-added and implemented getProjectById method

Deathroels committed:
BJS2-96054:
-added @Builder on Dto's class for future tests
-change validation on Dto's

Deathroels committed:
BJS2-96054:
-add methods inn mapper, for update

Deathroels committed:
BJS2-96054:
-add interface for filters

Deathroels committed:
BJS2-96054:
-fix signature in controller and facade for getAllProjects method

Deathroels committed:
BJS2-96054:
-fix create method
-add update, getAllProjects method
-need to impl deleteProjects, getProjectsByFilter methods

Deathroels committed:
BJS2-96054:
-add filters for projects
-add validation of project access

Deathroels committed:
BJS2-96054:
-added package for tests and blank tests for projects feature

Deathroels committed:
BJS2-96054:
-added tests too filters

Deathroels committed:
BJS2-96054:
-added tests too service

Scherbi1 commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.function.Consumer;
+
+@Mapper(componentModel = "spring", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
```
unmappedTargetPolicy = IGNORE
можешь юзать еще такую настройку в добавок
узнай, что делает

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,240 @@
+        projectId = 999L;
+        when(projectRepository.findById(projectId)).thenReturn(Optional.empty());
+
+        // When & Then
```
это кому?

Deathroels committed:
BJS2-96054:
-delete comment

Deathroels committed:
BJS2-96054:
-change conf of mapper annotation,set ReportingPolicy.IGNORE

Scherbi1 commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,30 @@
+    ProjectDto toProjectDto(Project project);
+
+    static void updateProjectFields(Project project, ProjectUpdateDto dto) {
+        updateIfNotNull(dto.name(), project::setName);
+        updateIfNotNull(dto.description(), project::setDescription);
```
а если в дто поле name и description придут одни пробелы?  по типу = "      "

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectFilterDto.java:
```diff
@@ -0,0 +1,15 @@
+
+@Builder
+public record ProjectFilterDto(
+        @Nullable
+        String name,
+
+        @Nullable
+        ProjectStatus status
```
почему оставляешь пробелы?

Deathroels replied:
если ты про тот что после status идет, то мне кажется так выглядит более отдленно, но если это редфлаг, могу убрать :)

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectUpdateDto.java:
```diff
@@ -0,0 +1,22 @@
+        ProjectStatus status,
+
+        @Nullable
+        ProjectVisibility visibility
```
.

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectDto.java:
```diff
@@ -0,0 +1,20 @@
+
+@Builder
+public record ProjectDto(
+        Long id,
+        String name,
+        String description,
+        Long ownerId,
+        LocalDateTime createdAt,
+        LocalDateTime updatedAt,
+        ProjectStatus status,
+        ProjectVisibility visibility
```
а тут не оставил.. в едином стиле делай

Deathroels replied:
и тут пробел после visibility, что не  так то, скобочка на след. строку уезжает и фсо,так же как и в каждом моем дто

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectDto.java:
```diff
@@ -0,0 +1,20 @@
+        String name,
+        String description,
+        Long ownerId,
+        LocalDateTime createdAt,
+        LocalDateTime updatedAt,
```
то есть LocalDateTime createdAt  и  LocalDateTime updatedAt должны возвращаться пользователю? и при создании проекта updatedAt  тоже будет указываться?

Deathroels replied:
Возможно это можно сделать отдельным, но да, я бы хотел отдавать пользователю при создании проекта данные, которые он будет видеть т.е. 
когда проект был создан и когда обновлен, думаю эти поля как раз для пользователя полезны в первую очередь

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectCreateDto.java:
```diff
@@ -0,0 +1,18 @@
+        String description,
+
+        @NotNull
+        ProjectVisibility visibility
```
.

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/project/ProjectCreateDto.java:
```diff
@@ -0,0 +1,18 @@
+
+@Builder
+public record ProjectCreateDto(
+        @NotNull
```
@NotBlank интересная аннотация есть) прочитай про нее

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+                .createdAt(LocalDateTime.now())
+                .updatedAt(LocalDateTime.now())
+                .visibility(projectCreateDto.visibility())
+                    .build();
```
почему маппишь тут?

Scherbi1 replied:
Желательно маппить в одном месте

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+        }
+        if (projectCreateDto.description().isEmpty() || projectCreateDto.description().isBlank()) {
+            throw new IllegalArgumentException("Project description cannot be empty");
+        }
```
у тебя есть аннотации для этого
И если пишешь ошибки, то используй String.format и указывай, где конкретно ошибка
Больше конкретики

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+        Project projectToDelete = filterProjectsByAccess(projectRepository.findAll().stream())
+                .filter(project -> project.getId().equals(projectId))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Project not found"));
```
то есть любой желающий может проект удалить?

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+
+        if (!Objects.equals(userId, project.getOwnerId())) {
+            throw new IllegalArgumentException("You can only update your own projects");
+        }
```
в валидатор или  в отдельный метод

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+                .filter(project -> Objects.equals(project.getOwnerId(), userId))
+                .forEach(project -> {
+                    if (Objects.equals(project.getName(), projectCreateDto.name())) {
+                        throw new IllegalArgumentException("Project with the same name already exists for this user");
```
Мне не очень нравится IllegalArgumentException, лучше кастомные писать 
Как было в User_Service
Но тут твое дело, дело вкуса короче

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,111 @@
+    public List<Project> getProjectsByFilter(ProjectFilterDto projectFilterDto) {
+        Stream<Project> projects = projectRepository.findAll().stream();
+
+        for (FilterProject filter : filters) {
+            if (filter.isApplication(projectFilterDto)) {
+                projects = filter.apply(projects, projectFilterDto);
+            }
+        }
```
я помню делал так же, но попробуй придумать что-нибудь, чтобы не переприсваивать стримы

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+                .visibility(ProjectVisibility.PUBLIC)
+                .build();
+
+        projectUpdateDto = new ProjectUpdateDto("Updated Name", "Updated Description", ProjectStatus.IN_PROGRESS, ProjectVisibility.PRIVATE);
```
слишком длинное сообщение, перенеси на другую строчку

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    @Test
+    void createProject_WithValidData_ShouldCreateProject() {
+        projectCreateDto = new ProjectCreateDto("Test create Project", "Test Description", ProjectVisibility.PUBLIC);
```
.

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    @Test
+    void createProject_WhenProjectWithSameNameExists_ShouldThrowException() {
+        projectCreateDto = new ProjectCreateDto("Test Project", "Test Description", ProjectVisibility.PUBLIC);
```
.

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    @Test
+    void createProject_WithEmptyName_ShouldThrowException() {
+        ProjectCreateDto createDto = new ProjectCreateDto("", "Description", ProjectVisibility.PUBLIC);
```
.

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    @Test
+    void createProject_WithBlankName_ShouldThrowException() {
+        ProjectCreateDto createDto = new ProjectCreateDto("   ", "Description", ProjectVisibility.PUBLIC);
```
.

Deathroels committed:
BJS2-96054:
-change and add more valuable annotation for validation

Deathroels committed:
BJS2-96054:
-add custom exception
-move mapping to mapper
-fix tests

Deathroels committed:
BJS2-96054:
-refactor imports and length of strings in test

Deathroels committed:
BJS2-96054:
-delete duplicate endpoint
-add ProjectValidator and moved some logic to him

Deathroels committed:
BJS2-96054:
-remove unused imports and a bit refactor

Deathroels committed:
BJS2-96054:
-add mock to test
-change access method in service
-delete wrong fields in mapper

Deathroels commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Test
+    void createProject_WhenProjectWithSameNameExists_ShouldThrowException() {
```
asd

Deathroels committed:
BJS2-96054:
-fixes

Deathroels committed:
BJS2-96054:
-refactor filter folder, move validate in private method

Deathroels committed:
BJS2-96054:
-fix tests

Deathroels committed:
Merge branch 'cerberus-master-stream12' into feature/projects-BJS2-96054

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,208 @@
+        userId = 1L;
+        projectId = 1L;
+
+        projectFirst = Project.builder()
+                .id(1L)
+                .name("Test Project")
+                .description("Test Description")
+                .ownerId(1L)
+                .status(ProjectStatus.CREATED)
+                .visibility(ProjectVisibility.PUBLIC)
+                .build();
+
+        projectSecond = Project.builder()
+                .id(2L)
+                .name("Test Project2")
+                .description("Test Description2")
+                .ownerId(2L)
+                .status(ProjectStatus.CREATED)
+                .visibility(ProjectVisibility.PUBLIC)
+                .build();
```
ты уверен, что так много переменных надо задавать в бефорИч?

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,13 @@
+public class ProjectValidator {
+    public static void validateProjectOwner(long userId, Project project) {
+        if (!Objects.equals(userId, project.getOwnerId())) {
+            throw new IllegalArgumentException("You can only update your own projects");
```
Да почему Illegal любите юзать(
использовали бы кастомные ошибки((

20cognosce replied:
Согласен

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+@RequiredArgsConstructor
+@Service
```
я бы еще логи добавил

20cognosce commented on src/main/java/faang/school/projectservice/dto/project/ProjectUpdateDto.java:
```diff
@@ -0,0 +1,23 @@
+import lombok.Builder;
+
+@Builder
+public record  ProjectUpdateDto(
```
пробел лишний

20cognosce commented on src/main/java/faang/school/projectservice/dto/project/ProjectUpdateDto.java:
```diff
@@ -0,0 +1,23 @@
+
+@Builder
+public record  ProjectUpdateDto(
+        @Nullable @NotBlank
+        String name,
+
+        @Nullable @NotBlank
+        String description,
```
они работают вместе? кажется они друг другу противоречат

20cognosce commented on src/test/java/faang/school/projectservice/service/project/filter/FilterProjectNameTest.java:
```diff
@@ -0,0 +1,65 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class FilterProjectNameTest {
```
Тесты на фильтры писать не надо

20cognosce commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,208 @@
+
+        assertNotNull(result);
+        verify(projectRepository).save(projectFirst);
+        assertEquals(projectUpdateDto.name(), result.getName());
+        assertEquals(projectUpdateDto.description(), result.getDescription());
+        assertEquals(projectUpdateDto.status(), result.getStatus());
+        assertEquals(projectUpdateDto.visibility(), result.getVisibility());
```
добавь ассерты перед выполнением действия, проверка исходного состояния
и проверь поля, которые не должны были обновиться тоже

20cognosce commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,100 @@
+    private final List<FilterProject> filters;
+
+    public Project createProject(ProjectCreateDto projectCreateDto) {
+        long userId = userContext.getUserId();
+
+        validateDuplicateProjectName(userId, projectCreateDto);
```
Ну вообще это косяк, ты никак не провалидировал юзера через user_service
Он существует вообще?

20cognosce reviewed (approved):
В целом всё чисто, принято

По возможности сделай интеграцию с юзер сервисом и проверяй юзера на статус, что не забанен, не удален, что он существует
